spec/unit/Java.spec.js

```diff
@@ -0,0 +1,158 @@
+    const Java = rewire('../../bin/templates/cordova/lib/Java');
+    let originalEnv = null;
+
+    beforeAll(() => {
+        originalEnv = Object.assign({}, process.env);
+    });
+
+    afterAll(() => {
+        process.env = Object.assign({}, originalEnv);
+    });
+
+    describe('getVersion', () => {
+        beforeEach(() => {
+            process.env = Object.assign({}, originalEnv);
+        });
+
+        it('runs', async () => {
+            Java.__set__('execa', () => Promise.resolve({
+                all: 'javac 1.8.0_275'
+            }));
+
+            const versionObj = {
+                major: 1,
+                minor: 8,
+                patch: 0,
+                build: 275,
+                string: '1.8.0_275'
+            };
+
+            await expectAsync(Java.getVersion()).toBeResolvedTo(versionObj);
+        });
+
+        it('produces a CordovaError on error', async () => {
+            Java.__set__('execa', () => Promise.reject({
+                shortMessage: 'test error'
+            }));
+            const emitSpy = jasmine.createSpy('events.emit');
+            Java.__set__('events', {
+                emit: emitSpy
+            });
+
+            await expectAsync(Java.getVersion()).toBeRejectedWithError(CordovaError, /Failed to run "javac -version"/);
+            expect(emitSpy).toHaveBeenCalledWith('verbose', 'test error');
+        });
+    });
+
+    describe('_ensure', () => {
+        beforeEach(() => {
+            process.env = Object.assign({}, originalEnv);
+            Java.__set__('javaIsEnsured', false);
+        });
+
+        it('with JAVA_HOME / without javac', async () => {
+            spyOn(utils, 'forgivingWhichSync').and.returnValue('');
+            process.env.JAVA_HOME = '/tmp/jdk';
+
+            await Java._ensure();
+
+            expect(process.env.PATH.split(path.delimiter))
+                .toContain(path.join(process.env.JAVA_HOME, 'bin'));
+        });
+
+        it('without JAVA_HOME / with javac - Mac OS X - success', async () => {
+            spyOn(utils, 'forgivingWhichSync').and.returnValue('/tmp/jdk/bin');
+            process.env.JAVA_HOME = '';
+            const fsSpy = jasmine.createSpy('fs').and.returnValue(true);
+            Java.__set__('fs', {
+                existsSync: fsSpy
+            });
+            Java.__set__('execa', jasmine.createSpy('execa').and.returnValue(Promise.resolve({
+                stdout: '/tmp/jdk'
+            })));
+
+            await Java._ensure();
+
+            expect(fsSpy).toHaveBeenCalledWith('/usr/libexec/java_home');
+            expect(process.env.JAVA_HOME).toBe('/tmp/jdk');
+        });
+
+        it('without JAVA_HOME / with javac - Mac OS X - error', async () => {
+            spyOn(utils, 'forgivingWhichSync').and.returnValue('/tmp/jdk/bin');
+            process.env.JAVA_HOME = '';
+            Java.__set__('fs', {
+                existsSync: jasmine.createSpy('fs').and.returnValue(true)
+            });
```

Comment:
Could also be simplified by using plain functions:
```suggestion
            Java.__set__('fs', { existsSync: () => true });
```